x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/adaptiveallocations/AdaptiveAllocationsScalerService.java

```diff
@@ -0,0 +1,332 @@
+                Stats nextStats = new Stats(
+                    nodeStats.getInferenceCount().orElse(0L),
+                    nodeStats.getPendingCount() == null ? 0 : nodeStats.getPendingCount(),
+                    nodeStats.getErrorCount() + nodeStats.getTimeoutCount() + nodeStats.getRejectedExecutionCount(),
+                    nodeStats.getAvgInferenceTime().orElse(0.0) / 1000.0
+                );
+                deploymentStats.put(nodeId, nextStats);
+                if (lastStats != null) {
+                    Stats recentStats = nextStats.sub(lastStats);
+                    recentStatsByDeployment.compute(
+                        assignmentStats.getDeploymentId(),
+                        (key, value) -> value == null ? recentStats : value.add(recentStats)
+                    );
+                }
+            }
+        }
+
+        if (statsTimeInterval == null) {
+            return;
+        }
+
+        for (Map.Entry<String, Stats> deploymentAndStats : recentStatsByDeployment.entrySet()) {
+            String deploymentId = deploymentAndStats.getKey();
+            Stats stats = deploymentAndStats.getValue();
+            AdaptiveAllocationsScaler adaptiveAllocationsScaler = scalers.get(deploymentId);
+            adaptiveAllocationsScaler.process(stats, statsTimeInterval, numberOfAllocations.get(deploymentId));
+            Integer newNumberOfAllocations = adaptiveAllocationsScaler.scale();
+            if (newNumberOfAllocations != null) {
+                Long lastScaleUpTimeMillis = lastScaleUpTimesMillis.get(deploymentId);
+                if (newNumberOfAllocations < numberOfAllocations.get(deploymentId)
+                    && lastScaleUpTimeMillis != null
+                    && now < lastScaleUpTimeMillis + SCALE_UP_COOLDOWN_TIME_MILLIS) {
+                    logger.debug("adaptive allocations scaler: skipping scaling down [{}] because of recent scaleup.", deploymentId);
+                    continue;
+                }
+                if (newNumberOfAllocations > numberOfAllocations.get(deploymentId)) {
+                    lastScaleUpTimesMillis.put(deploymentId, now);
+                }
+                UpdateTrainedModelDeploymentAction.Request updateRequest = new UpdateTrainedModelDeploymentAction.Request(deploymentId);
+                updateRequest.setNumberOfAllocations(newNumberOfAllocations);
+                ClientHelper.executeAsyncWithOrigin(
+                    client,
+                    ClientHelper.ML_ORIGIN,
+                    UpdateTrainedModelDeploymentAction.INSTANCE,
+                    updateRequest,
+                    ActionListener.wrap(updateResponse -> {
+                        logger.info("adaptive allocations scaler: scaled [{}] to [{}] allocations.", deploymentId, newNumberOfAllocations);
+                        threadPool.executor(MachineLearning.UTILITY_THREAD_POOL_NAME)
+                            .execute(
+                                () -> systemAuditor.info(
```

Comment:
I suggested using the `SystemAuditor` here but actually I think the `InferenceAuditor` auditor would make more sense as this is a update to a single deployment rather then the entire system.

The difference with the `InferenceAuditor` is that the `info(), warn() & error()` functions all take a `resourceId` parameter (in this case that is the deployment Id) the UI can then group all the notifications for a particular deployment.